File: app/src/main/java/com/synapse/social/studioasinc/NewGroupActivity.java
package com.synapse.social.studioasinc;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.synapse.social.studioasinc.model.User;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NewGroupActivity extends AppCompatActivity {

    private MaterialToolbar toolbar;
    private ImageView groupIcon;
    private EditText groupName;
    private LinearLayout disappearingMessagesLayout;
    private LinearLayout groupPermissionsLayout;
    private TextView membersCount;
    private RecyclerView membersRecyclerView;
    private FloatingActionButton fabCreateGroup;

    private ArrayList<User> selectedMembers = new ArrayList<>();
    private MembersAdapter membersAdapter;

    private FirebaseDatabase _firebase = FirebaseDatabase.getInstance();
    private DatabaseReference groupsRef = _firebase.getReference("groups");

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_new_group);

        toolbar = findViewById(R.id.toolbar);
        groupIcon = findViewById(R.id.group_icon);
        groupName = findViewById(R.id.group_name);
        disappearingMessagesLayout = findViewById(R.id.disappearing_messages_layout);
        groupPermissionsLayout = findViewById(R.id.group_permissions_layout);
        membersCount = findViewById(R.id.members_count);
        membersRecyclerView = findViewById(R.id.members_recyclerview);
        fabCreateGroup = findViewById(R.id.fab_create_group);

        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        toolbar.setNavigationOnClickListener(v -> onBackPressed());

        selectedMembers = (ArrayList<User>) getIntent().getSerializableExtra("selected_members");
        if (selectedMembers == null) {
            selectedMembers = new ArrayList<>();
        }

        setupRecyclerView();

        membersCount.setText("Members: " + selectedMembers.size());

        groupPermissionsLayout.setOnClickListener(v -> {
            Intent intent = new Intent(NewGroupActivity.this, GroupPermissionsActivity.class);
            startActivity(intent);
        });

        fabCreateGroup.setOnClickListener(v -> createGroup());
    }

    private void setupRecyclerView() {
        membersRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        membersAdapter = new MembersAdapter(selectedMembers);
        membersRecyclerView.setAdapter(membersAdapter);
    }

    private void createGroup() {
        String name = groupName.getText().toString().trim();
        if (TextUtils.isEmpty(name)) {
            Toast.makeText(this, "Group name cannot be empty", Toast.LENGTH_SHORT).show();
            return;
        }

        String groupId = groupsRef.push().getKey();
        if (groupId == null) {
            Toast.makeText(this, "Failed to create group", Toast.LENGTH_SHORT).show();
            return;
        }

        Map<String, Object> groupData = new HashMap<>();
        groupData.put("name", name);
        groupData.put("creatorId", FirebaseAuth.getInstance().getUid());
        groupData.put("createdAt", System.currentTimeMillis());

        Map<String, Boolean> members = new HashMap<>();
        members.put(FirebaseAuth.getInstance().getUid(), true);
        for (User user : selectedMembers) {
            members.put(user.getUid(), true);
        }
        groupData.put("members", members);

        groupsRef.child(groupId).setValue(groupData).addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                Toast.makeText(this, "Group created successfully", Toast.LENGTH_SHORT).show();
                Intent intent = new Intent(NewGroupActivity.this, ChatActivity.class);
                intent.putExtra("uid", groupId);
                intent.putExtra("isGroup", true);
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(intent);
                finish();
            } else {
                Toast.makeText(this, "Failed to create group", Toast.LENGTH_SHORT).show();
            }
        });
    }

    class MembersAdapter extends RecyclerView.Adapter<MembersAdapter.MemberViewHolder> {
        private List<User> members;

        public MembersAdapter(List<User> members) {
            this.members = members;
        }

        @NonNull
        @Override
        public MemberViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.synapse_users_list_cv, parent, false);
            return new MemberViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull MemberViewHolder holder, int position) {
            User user = members.get(position);
            holder.username.setText(user.getNickname() != null ? user.getNickname() : user.getUsername());
            Glide.with(holder.itemView.getContext()).load(user.getAvatar()).placeholder(R.drawable.avatar).into(holder.avatar);
        }

        @Override
        public int getItemCount() {
            return members.size();
        }

        class MemberViewHolder extends RecyclerView.ViewHolder {
            ImageView avatar;
            TextView username;

            public MemberViewHolder(@NonNull View itemView) {
                super(itemView);
                avatar = itemView.findViewById(R.id.profileCardImage);
                username = itemView.findViewById(R.id.username);
            }
        }
    }
}

File: app/src/main/java/com/synapse/social/studioasinc/NewGroupActivity.kt
package com.synapse.social.studioasinc

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.SearchView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.synapse.social.studioasinc.backend.SupabaseDatabaseService
import com.synapse.social.studioasinc.backend.interfaces.IDataListener
import com.synapse.social.studioasinc.backend.interfaces.IDataSnapshot
import com.synapse.social.studioasinc.backend.interfaces.IDatabaseError
import com.synapse.social.studioasinc.backend.interfaces.IDatabaseService
import com.synapse.social.studioasinc.model.User

class NewGroupActivity : AppCompatActivity() {

    private lateinit var usersRecyclerView: RecyclerView
    private lateinit var searchView: SearchView
    private lateinit var fabNext: FloatingActionButton

    private lateinit var usersAdapter: UsersAdapter
    private val usersList = mutableListOf<User>()
    private val selectedUsers = mutableListOf<String>()

    private val dbService: IDatabaseService by lazy { (application as SynapseApp).getDatabaseService() }
    private val database by lazy { dbService.getReference("skyline/users") }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_new_group)

        usersRecyclerView = findViewById(R.id.users_recycler_view)
        searchView = findViewById(R.id.search_view)
        fabNext = findViewById(R.id.fab_next)

        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = "Add Members"

        usersAdapter = UsersAdapter(usersList) { user, isChecked ->
            if (isChecked) {
                selectedUsers.add(user.uid)
            } else {
                selectedUsers.remove(user.uid)
            }
        }

        usersRecyclerView.apply {
            layoutManager = LinearLayoutManager(this@NewGroupActivity)
            adapter = usersAdapter
        }

        fabNext.setOnClickListener {
            if (selectedUsers.isNotEmpty()) {
                val intent = Intent(this, CreateGroupActivity::class.java)
                intent.putStringArrayListExtra("selected_users", ArrayList(selectedUsers))
                startActivity(intent)
            } else {
                Toast.makeText(this, "Select at least one member", Toast.LENGTH_SHORT).show()
            }
        }

        setupSearch()
        fetchUsers()
    }

    private fun setupSearch() {
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String?): Boolean {
                usersAdapter.filter(newText)
                return true
            }
        })
    }

    @Suppress("UNCHECKED_CAST")
    private fun fetchUsers() {
        dbService.getData(database, object : IDataListener {
            override fun onDataChange(dataSnapshot: IDataSnapshot) {
                usersList.clear()
                if (dataSnapshot.exists()) {
                    val userMaps = dataSnapshot.getValue(List::class.java) as? List<Map<String, Any>>
                    userMaps?.forEach { userMap ->
                        val user = User(
                            userMap["uid"] as? String ?: "",
                            userMap["username"] as? String ?: "",
                            userMap["nickname"] as? String ?: "",
                            userMap["avatar"] as? String ?: ""
                        )
                        usersList.add(user)
                    }
                }
                usersAdapter.notifyDataSetChanged()
            }

            override fun onCancelled(databaseError: IDatabaseError) {
                Toast.makeText(this@NewGroupActivity, "Failed to load users: ${databaseError.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }
}

class UsersAdapter(
    private var users: List<User>,
    private val onUserSelected: (User, Boolean) -> Unit
) : RecyclerView.Adapter<UsersAdapter.UserViewHolder>() {

    private var filteredUsers: MutableList<User> = users.toMutableList()

    override fun onCreateViewHolder(parent: android.view.ViewGroup, viewType: Int): UserViewHolder {
        val view = android.view.LayoutInflater.from(parent.context)
            .inflate(R.layout.item_user_selectable, parent, false)
        return UserViewHolder(view)
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        holder.bind(filteredUsers[position])
    }

    override fun getItemCount(): Int = filteredUsers.size

    fun filter(query: String?) {
        filteredUsers = if (query.isNullOrBlank()) {
            users.toMutableList()
        } else {
            users.filter {
                (it.username?.contains(query, ignoreCase = true) ?: false) ||
                (it.nickname?.contains(query, ignoreCase = true) ?: false)
            }.toMutableList()
        }
        notifyDataSetChanged()
    }

    inner class UserViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val userName: TextView = itemView.findViewById(R.id.user_name)
        private val userAvatar: ImageView = itemView.findViewById(R.id.user_avatar)
        private val userCheckbox: CheckBox = itemView.findViewById(R.id.user_checkbox)

        fun bind(user: User) {
            userName.text = if (!user.nickname.isNullOrEmpty()) user.nickname else user.username
            Glide.with(itemView.context).load(user.avatar).placeholder(R.drawable.avatar).into(userAvatar)

            itemView.setOnClickListener {
                userCheckbox.isChecked = !userCheckbox.isChecked
            }

            userCheckbox.setOnCheckedChangeListener { _, isChecked ->
                onUserSelected(user, isChecked)
            }
        }
    }
}